feat: 우선순위 스케줄링 구현

hdh853 opened the pull request:
## 🔗 관련 이슈
<!-- 이슈 번호를 연결해주세요 -->
closes #12 

## 📌 작업 내용
### 구현 사항
<!-- 구현내용에 대해서 설명해주세요. -->
<!-- 예시 :
- 이슈 템플릿, PR 템플릿 작성
- READEME 파일 최초 작성
 --> 
- ready_list, 세마포어의 wait_list에 들어갈 때 우선순위 반영되어 삽입
- cond_signal 에서 sema_up 하기 전 list_sort하는 로직 추가
- 우선순위 계산하기 위한 priority_greater, cond_priority_greater 함수 추가

### 변경된 파일
<!-- 구현 내용으로 인해 변경된 파일에 대해 설명해주세요 -->
<!-- 예시:
- README.md
  - readme 파일 최초 생성
- issue_template.md
  - 이슈 템플릿 최초 생성
-->
- thread.h
  -  priority_greater 함수 헤더 추가
  
- synch.c
  - sema_down시 우선순위 고려해 삽입하도록 list_push_back 에서 list_insert_ordered로 변경
  - sema_up시 우선순위 고려해 thread_yield()하는 로직 추가
  - cond_signal에서 sema_up하기 전 list_sort하는 로직 추가
  - cond에서 사용하는 우선순위 비교함수 cond_priority_greater 함수 추가
  
- thread.c
  - thread_create, thread_set_priority 후 우선순위 검사해 yield()하는 로직 추가
  - thread_unblock에서 ready_list에 넣을 때 list_insert_ordered로 넣도록 수정

### 추가 사항
<!-- 트러블 슈팅 등 추가로 설명할 내용이 있으시다면 아래에 작성해주세요. -->

## ✅ 셀프 체크리스트
<!-- 체크리스트를 모두 완료한 후 제출해주세요 -->
<!-- 완료 처리를 표시하려면 [x], 완료되지 않았다면 [ ] -->
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 주석이나 디버깅 코드를 제거했나요?
- [x] 관련 테스트가 모두 통과하나요?
- [x] 커밋 메시지가 컨벤션에 맞나요?

hdh853 committed:
style: synch.c 포멧 적용

hdh853 committed:
feat: priority scheduling 구현
- ready_list에 우선순위 고려해서 넣도록 수정
 - thread_create, set_thread_priority 시 우선순위 고려해서 yield 하도록 수정
- 우선순위 비교 위한 priority_greater 함수 생성
- 세마포어의 waiter리스트에 넣을 때도 우선순위 고려하도록 수정
- cond_signal시 정렬 후 sema_up 하도록 수정

hyunmuam commented on threads/synch.c:
```diff
@@ -105,12 +107,21 @@ void sema_up(struct semaphore* sema)
+            thread_yield();
+        }
+    }
 }
```
if문을 2개로 쪼개는 것을 고려해보는 것은 어떤가요?
가독성과 조건별 역할 구분이 명확해질 것 같습니다.
```c
if (wakeup_thread == NULL || intr_context() || old_level != INTR_ON)
    return;
    
if (thread_current()->priority < wakeup_thread->priority) {
    thread_yield();
}  
```
두번째로 `old_level == INTR_ON` 검사에 대한 의문이 있습니다.
의도는 인터럽트를 disable한 상황에서 양보하지 않기 위한 것으로 이해됩니다.
다만 `intr_context()` 검사와 중복되는 부분이 있는 것 같은데
혹시 이 조건이 필요한 구체적인 시나리오가 있을까요?

hdh853 replied:
그러네요 if문을 쪼개놓으니까 의도가 잘보이는것 같습니다. 피드백 감사합니다.
old_level == INTR_ON을 검사한 이유에 대해 말씀해주셨는데요, intr_context()와 검사하는 영역이 다르다 생각했습니다
우선 intr_context()를 검사한건 '지금 sema_up이 호출된 곳이 인터럽트 핸들러인가?' 를 검사하기 위해 사용했고,
old_level == INTR_ON을 검사한건 'sema_up이 호출된 시점이 원자성이 보장되어야 하는 시점인가?' 를 검사하기 위해 사용했습니다. 즉 호출 전 밖에서 인터럽트를 끄고 sema_up을 호출했다면, 내부에서 마음대로 yield()하면 원자성이 보장되지 않을것 같아서요.
 찾아보니 현재의  pintOS 소스에 그런식으로 sema_up을 사용하는 코드는 없긴 하네요.. 그래도 그런 경우가 생길수도 있을것 같아서 남겨두는게 좋을것 같은데 어떻게 생각하시나요?

hyunmuam replied:
@hdh853 
맞는 것 같습니다. 인터럽트를 비활성화한 상태(임계영역 내)에서 sema_up을 호출할 때 현재 스레드가 양보하지 않도록 `old_level == INTR_ON` 검사가 필요해 보입니다.
다만 이 조건의 의도가 코드상에서 바로 드러나지 않으니, 해당 검사가 왜 필요한지 주석으로 설명을 추가해 주시면 좋을 것 같습니다.

hyunmuam commented on threads/synch.c:
```diff
@@ -297,6 +308,8 @@ void cond_signal(struct condition* cond, struct lock* lock UNUSED)
     ASSERT(lock_held_by_current_thread(lock));
 
+    // 정렬 후 풀기
+    list_sort(&cond->waiters, cond_priority_greater, NULL);
```
`&cond->waiters`에서 꺼내는 부분은 `cond_signal` 밖에 없습니다.
그러면 매번 정렬하는 것보다는 최댓값을 가져오는게 더 효율적일 수 있을 것 같습니다.
매번 정렬하면 복잡도가 O(n log n)이고, 최댓값만 가져온다면 O(n)으로 줄일 수 있으니까요. `list_min()`/`list_max()` 사용을 고려해보시는 건 어떤가요?

hdh853 replied:
아 그런방법이 있었네요 정렬하고 뽑거나, 넣을때 정렬하거나 두가지 경우만 생각했었는데 min/max를 사용하는게 좀 더 효과적일것 같습니다. 피드백 감사합니다!

hyunmuam reviewed (changes requested):
destination branch를 develop으로 변경해주시면 감사하겠습니다.
(현재: main → 변경: develop)

+) PR올리면 리뷰어 지정도 해주시면 감사겠습니다. 리뷰어 지정을 안해주시면 저에게 알람이 안와요 ㅠㅠ
+) PR 타이틀도 PR내용에 맞게 변경해주세요~

hdh853 committed:
refactor: 우선순위 체크 방식 변경
-  sema : 넣을 때 정렬해서 넣는 방식에서 빼낼 때 list_min으로 빼내는 방식으로 수정
- cond : 빼낼때 sort 후 빼내던 방식에서 list_min으로 빼내는 방식으로 수정

hdh853 committed:
refactor: sema_up에서 if문 분기 및 주석 추가

hdh853 commented:
말씀해주신 피드백 반영 완료했습니다!
항상 양질의 피드백 감사합니다 ㅎㅎ

hyunmuam commented on threads/synch.c:
```diff
@@ -57,100 +59,111 @@ sema_init (struct semaphore *sema, unsigned value) {
+
+    old_level = intr_disable();
+    while (sema->value == 0) {
+        // list_insert_ordered(&sema->waiters, &thread_current()->elem, priority_greater, NULL);
```
@hdh853 
지워주세용~

hyunmuam commented on threads/synch.c:
```diff
@@ -295,16 +308,19 @@ cond_wait (struct condition *cond, struct lock *lock) {
+        struct list_elem* max_elem = list_min(&cond->waiters, cond_priority_greater, NULL);
+        list_remove(max_elem);
+        sema_up(&list_entry(max_elem, struct semaphore_elem, elem)->semaphore);
+    }
```
@hdh853 
주석이 설명하는 것과 코드가 동작하는게 일치하지 않는 것 같습니다.

hyunmuam reviewed (commented):
주석 관련되서 쬐끔만 수정하면 될 것 같습니다~

hdh853 committed:
refactor: 잘못된 주석 수정 및 제거

hyunmuam reviewed (approved):
수고하셨습니다.
